### PR TITLE
Added module to describe EC2 Placement Groups.

### PR DESCRIPTION
**Pull request by ghost**
##### SUMMARY
New module to return facts for Amazon EC2 Placement Groups.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
ec2_placement_group_facts

##### ANSIBLE VERSION
```
ansible 2.5.0 (ec2_placement_group_facts a1bad18c06) last updated 2017/11/23 11:36:49 (GMT +100)
  config file = None
  configured module search path = ['/Users/brad.macpherson/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/brad.macpherson/workspace/ansible/lib/ansible
  executable location = /Users/brad.macpherson/workspace/ansible/bin/ansible
  python version = 3.6.3 (default, Oct  4 2017, 06:09:38) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.37)]
```

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+                                      boto3_conn,
+                                      ec2_argument_spec,
+                                      get_aws_connection_info,
+                                      get_ec2_security_group_ids_from_names)
+
+
+def get_placement_groups_details(connection, module):
+    names = module.params.get("names")
+    if len(names) > 0:
+        response = connection.describe_placement_groups(
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py]: All AWS API calls should have exception handling

File: lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+'''
+
+try:
+    import boto3
+    from botocore.exceptions import ClientError
+    HAS_BOTO = True
+except ImportError:
+    HAS_BOTO = False
+
+from ansible.module_utils.basic import AnsibleModule
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py]: use AnsibleAWSModule now (see latest [guidelines](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/GUIDELINES.md) which also covers exception handling etc)

File: lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    if not HAS_BOTO:
+        module.fail_json(msg='boto required for this module')
+
+    region, ec2_url, aws_connect_params = get_aws_connection_info(
+        module, boto3=True)
+
+    if region:
+        try:
+            connection = boto3.client('ec2', region, **aws_connect_params)
+        except (boto3.exceptions.NoAuthHandlerFound, AnsibleAWSError) as e:
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py]: None of these errors ever happen. Please update to the latest AWS connection guidelines.

File: lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+            module.fail_json(msg=str(e))
+    else:
+        module.fail_json(msg="region must be specified")
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py]: boto3_conn now copes with RegionNotFound errors for you

File: lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+      sample: "cluster"
+
+'''
+
+try:
+    import boto3
+    from botocore.exceptions import ClientError
+    HAS_BOTO = True
+except ImportError:
+    HAS_BOTO = False
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py]: HAS_BOTO3 errors are handled by AnsibleAWSModule (and before that, you could import it from module_utils.ec2).
No need to import boto3.

**Review by willthames — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py (hunk 4 above)
Comment (ghost) [reply lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py]: No worries - addressed (hopefully all) these in this module and in ec2_placement_groups.py (separate PR).

File: lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py (hunk 5 above)
Comment (ghost) [reply lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py]: Duly updated to import HAS_BOTO3.

File: lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py (hunk 2 above)
Comment (ghost) [reply lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py]: Done

File: lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py (hunk 1 above)
Comment (ghost) [reply lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py]: Fixed; one try block for both variants of the call.

File: lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py (hunk 3 above)
Comment (ghost) [reply lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py]: Done

**Review by willthames — APPROVED** [review]

File: lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
+#!/usr/bin/python
+#
+# This is a free software: you can redistribute it and/or modify
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py]: The license block is now much simpler: http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#copyright

File: lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+# Note: These examples do not set authentication details or the AWS region,
+# see the AWS Guide for details.
+
+# List all placement groups.
+- ec2_placement_group_facts:
+  register: all_ec2_placement_groups
+
+# List two placement groups.
+- ec2_placement_group_facts:
+  names:
```
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py]: This example isn't valid - `names` needs to be indented one more level.

File: lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+     - my-cluster
+     - my-other-cluster
+  register: specific_ec2_placement_groups
+
```
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py]: Also can you include an example of getting a specific-named placement group out of the list, like: 

```
- debug: msg="{{ specific_ec2_placement_groups | json_query(\"[?name=='my-cluster']\") }}"
```

**Review by ryansb — APPROVED** [review]
Code is :1st_place_medal:  - minor doc change requests then it's good to merge.

File: lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py (hunk 8 above)
Comment (willthames) [reply lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py]: @ryansb - you don't have to escape json_query that much, as backticks are valid string quotes

```
- debug: msg="{{ specific_ec2_placement_groups | json_query('[?name==`my-cluster`]') }}"
```

File: lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py (hunk 7 above)
Comment (ghost) [reply lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py]: Done

File: lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py (hunk 8 above)
Comment (ghost) [reply lib/ansible/modules/cloud/amazon/ec2_placement_group_facts.py]: Added

**Comment by s-hertel** [comment]
Thanks @iiibrad!

**Pull request merged**